x/dex/keeper/settlement.go

```diff
@@ -11,29 +11,67 @@ import (
 func (k Keeper) SetSettlements(ctx sdk.Context, contractAddr string, priceDenom string, assetDenom string, settlements types.Settlements) {
 	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.SettlementEntryPrefix(contractAddr, priceDenom, assetDenom))
 	for _, settlement := range settlements.GetEntries() {
-		existing, found := k.GetSettlementsState(ctx, contractAddr, priceDenom, assetDenom, settlement.OrderId)
+		existing, found := k.GetSettlementsState(ctx, contractAddr, priceDenom, assetDenom, settlement.Account, settlement.OrderId)
 		if found {
 			existing.Entries = append(existing.Entries, settlement)
 		} else {
 			existing = settlements
 		}
 
-		orderIDBytes := make([]byte, 8)
-		binary.BigEndian.PutUint64(orderIDBytes, settlement.OrderId)
 		b := k.Cdc.MustMarshal(&existing)
-		store.Set(orderIDBytes, b)
+		store.Set(getSettlementKey(settlement.OrderId, settlement.Account), b)
 	}
 }
 
-func (k Keeper) GetSettlementsState(ctx sdk.Context, contractAddr string, priceDenom string, assetDenom string, orderID uint64) (val types.Settlements, found bool) {
+func (k Keeper) GetSettlementsState(ctx sdk.Context, contractAddr string, priceDenom string, assetDenom string, account string, orderId uint64) (val types.Settlements, found bool) {
 	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.SettlementEntryPrefix(contractAddr, priceDenom, assetDenom))
-	orderIDBytes := make([]byte, 8)
-	binary.BigEndian.PutUint64(orderIDBytes, orderID)
-	b := store.Get(orderIDBytes)
+	b := store.Get(getSettlementKey(orderId, account))
 	val = types.Settlements{}
 	if b == nil {
 		return val, false
 	}
 	k.Cdc.MustUnmarshal(b, &val)
 	return val, true
 }
+
+func (k Keeper) GetSettlementsStateForAccount(ctx sdk.Context, contractAddr string, priceDenom string, assetDenom string, account string) []types.Settlements {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.SettlementEntryPrefix(contractAddr, priceDenom, assetDenom))
+	res := []types.Settlements{}
+	iterator := sdk.KVStorePrefixIterator(store, []byte(account))
+
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var val types.Settlements
+		k.Cdc.MustUnmarshal(iterator.Value(), &val)
+		res = append(res, val)
+	}
+
+	return res
+}
+
+func (k Keeper) GetAllSettlementsState(ctx sdk.Context, contractAddr string, priceDenom string, assetDenom string, limit int) []types.Settlements {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.SettlementEntryPrefix(contractAddr, priceDenom, assetDenom))
+	res := []types.Settlements{}
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
```

Comment:
is there other access way rather than iterating through the KVStore? like getting a slice with length n instead?

Reply:
i think the point of iterator is so that we don't have to fetch everything at once. Since it's a key-value store there isn't much of an N+1 concern